experimental/CliffordAlgebras/src/CliffordAlgebras.jl

```diff
@@ -0,0 +1,579 @@
+mutable struct CliffordAlgebraElem{T,S,P} <: Hecke.AbstractAssociativeAlgebraElem{T}
+  parent::P
+  coeffs::Vector{T}
+  even_coeffs::Vector{T}
+  odd_coeffs::Vector{T}
+
+  #Return the 0-element of the Clifford algebra C
+  function CliffordAlgebraElem{T,S,P}(C::CliffordAlgebra{T,S}) where {T,S,P}
+    newelt = new{T,S,CliffordAlgebra{T,S}}(C, fill(C.base_ring(), C.dim))
+    _set_even_odd_coeff!(newelt)
+    return newelt
+  end
+
+  CliffordAlgebraElem(C::CliffordAlgebra) =
+    CliffordAlgebraElem{elem_type(C.base_ring),typeof(C.gram),typeof(C)}(C)
+
+  #Return the element in the Clifford algebra C with coefficient vector coeff wrt. the canonical basis
+  function CliffordAlgebraElem{T,S,P}(
+    C::CliffordAlgebra{T,S}, coeff::Vector{R}
+  ) where {T,S,P,R<:FieldElem}
+    @req length(coeff) == C.dim "invalid length of coefficient vector"
+    newelt = new{T,S,CliffordAlgebra{T,S}}(C, coeff)
+    _set_even_odd_coeff!(newelt)
+    return newelt
+  end
+
+  CliffordAlgebraElem(C::CliffordAlgebra{T,S}, coeff::Vector{T}) where {T,S} =
+    CliffordAlgebraElem{elem_type(C.base_ring),typeof(C.gram),typeof(C)}(C, coeff)
+
+  function CliffordAlgebraElem(C::CliffordAlgebra{T,S}, coeff::Vector{R}) where {T,S,R}
+    K = C.base_ring
+    @req __can_convert_coeff(coeff, K) "entries of coefficient vector are not contained in $(K)"
+    return CliffordAlgebraElem{elem_type(C.base_ring),typeof(C.gram),typeof(C)}(
+      C, K.(coeff)
+    )
+  end
+end
+
+elem_type(::Type{CliffordAlgebra{T,S}}) where {T,S} =
+  CliffordAlgebraElem{T,S,CliffordAlgebra{T,S}}
+
+parent_type(::Type{CliffordAlgebraElem{T,S,P}}) where {T,S,P} = P
+
+base_ring_type(C::CliffordAlgebra{T,S}) where {T,S} = typeof(base_ring(C))
+
+################################################################################
+#
+#  Construction
+#
+################################################################################
+
+### Algebra ###
+
+@doc raw"""
+    clifford_algebra(qs::QuadSpace) -> CliffordAlgebra
+
+Return the Clifford algebra of the quadratic space 'qs'.
+"""
+clifford_algebra(qs::Hecke.QuadSpace) =
+  CliffordAlgebra{elem_type(base_ring(qs)),typeof(gram_matrix(qs))}(qs)
+
+### Elements ###
+(C::CliffordAlgebra)() = CliffordAlgebraElem(C)
+
+function (C::CliffordAlgebra)(a::R) where {R<:RingElem}
+  res = fill(zero(a), dim(C))
+  res[1] = a
+  return CliffordAlgebraElem(C, res)
+end
+
+function (C::CliffordAlgebra)(a::R) where {R<:Number}
+  res = fill(zero(a), dim(C))
+  res[1] = a
+  return CliffordAlgebraElem(C, res)
+end
+
+(C::CliffordAlgebra)(coeff::Vector{R}) where {R} = CliffordAlgebraElem(C, coeff)
+
+################################################################################
+#
+#  Basic field access
+#
+################################################################################
+
+### Algebra ###
+@doc raw"""
+    base_ring(C::CliffordAlgebra) -> Ring
+
+Return the base ring of the Clifford algebra $C$.
+"""
+base_ring(C::CliffordAlgebra) = C.base_ring
```

Comment:
```suggestion
base_ring(C::CliffordAlgebra) = C.base_ring::TypeAnnotation
```
this needs a type annotation, to not infer `Ring` here. You should be able to just put your updated implementation of `base_ring_type` there

Reply:
i.e.
```suggestion
base_ring(C::CliffordAlgebra) = C.base_ring::base_ring_type(typeof(C))
```

Reply:
Fixed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/73f677cfe5c0362d5f9763d3d3b24a6e832e63a2)